pkg/infra/BlockCollector.go

```diff
@@ -0,0 +1,42 @@
+package infra
+
+import "sync"
+
+type BlockCollector struct {
+	sync.Mutex
+	Observers int
+	Registry  map[uint64]int
+}
+
+func NewBlockCollector(observers int) *BlockCollector {
+	registry := make(map[uint64]int)
+	return &BlockCollector{
+		Observers: observers,
+		Registry:  registry,
+	}
+}
+
+func (bc *BlockCollector) Commit(block uint64) (committed bool) {
+	if bc.Observers == 1 {
+		return true
+	}
+	bc.Lock()
+	defer bc.Unlock()
+	// read map
+	if i, ok := bc.Registry[block]; ok {
+		if i > 0 {
+			i = i - 1
+			if i == 0 {
+				//remove from map
+				delete(bc.Registry, block)
+				return true
+			}
+			// update map
+			bc.Registry[block] = i
+		}
+	} else {
+		// write map
+		bc.Registry[block] = bc.Observers - 1
+	}
+	return false
+}
```

Comment:
```go
package infra

import "sync"

type BlockCollector struct {
	sync.Mutex

	threshold int
	registry  map[uint64]int
}

func NewBlockCollector(threshold int) *BlockCollector {
	registry := make(map[uint64]int)
	return &BlockCollector{
		threshold: threshold,
		registry:  registry,
	}
}

func (bc *BlockCollector) Commit(block uint64) (committed bool) {
	if bc.threshold == 1 {
		return true
	}

	bc.Lock()
	defer bc.Unlock()

	remaining, ok := bc.registry[block]
	if !ok {
		bc.registry[block] = bc.threshold - 1
		return false
	}

	if remaining == 1 {
		delete(bc.registry, block)
		return true
	}

	bc.registry[block]--
	return false
}
```
- no need to export internal members
- less nested if-else
- more meaningful var names

Reply:
thx